File: Widen/LC646_Maximum_Length_of_Pair_Chain.py
"""
646. Maximum Length of Pair Chain
You are given n pairs of numbers. In every pair, the first number is always smaller than the second number.

Now, we define a pair (c, d) can follow another pair (a, b) if and only if b < c. Chain of pairs can be formed in this fashion.

Given a set of pairs, find the length longest chain which can be formed. You needn't use up all the given pairs. You can select pairs in any order.

Example 1:
Input: [[1,2], [2,3], [3,4]]
Output: 2
Explanation: The longest chain is [1,2] -> [3,4]
"""

# turns out this is a greedy problem, sorted by the second parameter!
# time complexity -- O(NlogN)
# a good example as comparsion between greedy and dp
# Runtime: 204 ms, faster than 97.11% of Python3 online submissions for Maximum Length of Pair Chain.
# Memory Usage: 14.1 MB, less than 7.14% of Python3 online submissions for Maximum Length of Pair Chain.
class Solution:
    def findLongestChain(self, pairs: List[List[int]]) -> int:
        res = 0
        curr = float("-inf")
        for x, y in sorted(pairs, key=lambda x:x[1]):
            if curr >= x:
                continue
            curr = y
            res += 1
        return res
        


# time complexity -- O(N^2 + NlogN)
# Runtime: 3992 ms, faster than 7.44% of Python3 online submissions for Maximum Length of Pair Chain.
# Memory Usage: 14.2 MB, less than 7.14% of Python3 online submissions for Maximum Length of Pair Chain.
class Solution:
    def findLongestChain(self, pairs: List[List[int]]) -> int:
        N = len(pairs)
        self.pairs = sorted(pairs, key = lambda x:x[0])
        self.visited = [False for _ in range(N)]
        self.lens = [1 for _ in range(N)]
        res = -1
        for i in range(N):
            res = max(res, self.dfs(i, N))
        return res
    
    def dfs(self, start_idx, N):
        if self.visited[start_idx]:
            return self.lens[start_idx]
        self.visited[start_idx] = True
        if start_idx == N - 1:
            return 1
        base_len = 0
        for i in range(start_idx, N):
            if self.pairs[i][0] <= self.pairs[start_idx][1]:
                continue
            tmp_max_len = self.dfs(i, N)
            base_len = max(base_len, tmp_max_len)
        self.lens[start_idx] = base_len + 1
        return self.lens[start_idx]
    
            
        

File: Widen/LC646_Maximum_Length_of_Pair_Chain.go
/*
646. Maximum Length of Pair Chain
You are given n pairs of numbers. In every pair, the first number is always smaller than the second number.

Now, we define a pair (c, d) can follow another pair (a, b) if and only if b < c. Chain of pairs can be formed in this fashion.

Given a set of pairs, find the length longest chain which can be formed. You needn't use up all the given pairs. You can select pairs in any order.

Example 1:
Input: [[1,2], [2,3], [3,4]]
Output: 2
Explanation: The longest chain is [1,2] -> [3,4]
*/


type Pair struct {x, y int}

type PairSorter []Pair

func (ps PairSorter) Len() int{
    return len(ps)
}

func (ps PairSorter) Swap(i, j int){
    ps[i], ps[j] = ps[j], ps[i]
}

func (ps PairSorter) Less(i, j int) bool{
    return ps[i].y < ps[j].y
}

func findLongestChain(pairs [][]int) int {
    pair_arr := make([]Pair, 0, len(pairs))
    for _, p := range pairs{
        pair_arr = append(pair_arr, Pair{p[0], p[1]})
    }
    ps := PairSorter(pair_arr)
    sort.Sort(ps)
    res := 0
    curr := math.MinInt32
    for _, p := range ps{
        if curr >= p.x{
            continue
        } else {
            curr = p.y
            res += 1
        }
    }
    return res
}